prov/gni/include/gnix_buddy_allocator.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2015 Los Alamos National Security, LLC. All rights reserved.
+ * Copyright (c) 2015 Cray Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef _GNIX_BUDDY_ALLOCATOR_H_
+#define _GNIX_BUDDY_ALLOCATOR_H_
+
+#include "fi_list.h"
+#include "gnix_bitmap.h"
+#include "gnix_util.h"
+#include <math.h>
+#include <stdlib.h>
+
+#define MIN_BLOCK_SIZE 16
+
+/* evaluates to zero if X is not a power of two, otherwise evaluates to X - 1 */
+#define IS_NOT_POW_TWO(X) (((X) & (~(X) + 1)) ^ (X))
+
+/* Find the block size (in bytes) required for allocating LEN bytes */
+#define BLOCK_SIZE(LEN, MIN) ((LEN) <= (MIN) ? (MIN) :\
+			      (IS_NOT_POW_TWO(LEN)) ? (((LEN) << 1) & ~(LEN)) :\
+			      (LEN))
+
+/* Find the bitmap index for block X */
+#define BITMAP_INDEX(X, BASE, MIN, LEN) (size_t) ((size_t) ((X) - (BASE)) /\
+					 (MIN) + 2 * log2((LEN) / (MIN)))
+
+/*
+ * The following macro doesn't work when the base address starts at zero.
+ * #define BUDDY(X, SIZE_X, BASE) (void *) ((size_t) (X) ^ (SIZE_X))
+ */
+
+/* Find the address of X's buddy block:
+ * If the "index" of block X is even then the buddy must be to the right of X,
+ * otherwise the buddy is to the left of X.
+ */
+#define BUDDY(X, LEN, BASE) (void *) ((((size_t) (BASE) - (size_t) (X)) /\
+				       (LEN)) % 2 ? (size_t) (X) - (LEN) :\
+				      (size_t) (X) + (LEN))
+
+/* Calculate the offset of a free block, OFFSET = MIN * 2^MULT. */
+#define OFFSET(MIN, MULT) ((MIN) * (1 << (MULT)))
+
+/* Find the index into the free list with block size LEN. */
+#define LIST_INDEX(LEN, MIN) (size_t) (log2((LEN) / (double) (MIN)))
```

Comment:
That makes sense.  Great.

> On Jan 8, 2016, at 11:26 AM, James Swaro <notifications@github.com> wrote:
> 
> In prov/gni/include/gnix_buddy_allocator.h:
> 
> > + * #define BUDDY(X, SIZE_X, BASE) (void *) ((size_t) (X) ^ (SIZE_X))
> > + */
> > +
> > +/* Find the address of X's buddy block:
> > + * If the "index" of block X is even then the buddy must be to the right of X,
> > + * otherwise the buddy is to the left of X.
> > + */
> > +#define BUDDY(X, LEN, BASE) (void *) ((((size_t) (BASE) - (size_t) (X)) /\
> > +				       (LEN)) % 2 ? (size_t) (X) - (LEN) :\
> > +				      (size_t) (X) + (LEN))
> > +
> > +/* Calculate the offset of a free block, OFFSET = MIN * 2^MULT. */
> > +#define OFFSET(MIN, MULT) ((MIN) * (1 << (MULT)))
> > +
> > +/* Find the index into the free list with block size LEN. */
> > +#define LIST_INDEX(LEN, MIN) (size_t) (log2((LEN) / (double) (MIN)))
> Your len will never be over 4GB, so I don't see why the uint32_t version would not work.
> 
> —
> Reply to this email directly or view it on GitHub.
>